ceph_iscsi_config/device_status.py

```diff
@@ -0,0 +1,165 @@
+        if stat_cnt.name == "cmd_timed_out_cnt" or \
+           stat_cnt.name == "conn_lost_cnt":
+            if abs(stat_cnt.cnt - stat_cnt.last_cnt) >= 1:
+                self.state = "Degraded - cluster access failure"
+                self.changed_state = True
+                self.stable_cnt = 0
+                return
+
+        if stat_cnt.name == "lock_lost_cnt" and \
+           abs(stat_cnt.cnt - stat_cnt.last_cnt) >= \
+           settings.config.lock_lost_cnt_thresh:
+            self.state = "Degraded - excessive failovers"
+            self.changed_state = True
+            self.stable_cnt = 0
+            return
+
+    def update_status(self, gw, status, update_lock_owner=True):
+        if status is None:
+            # Sometimes status calls will return empty statuses even though
+            # there is valid data. We might not see it until the Nth call.
+            return
+
+        counter_dict = self.gw_counter_lookup.get(gw)
+        if counter_dict is None:
+            counter_dict = {}
+
+        for name, val in status.items():
+            if update_lock_owner is True and name == "lock_owner" and val == "true":
+                print("gw = {}".format(gw))
+                self.lock_owner = gw
+                continue
+
+            if name != "cmd_timed_out_cnt" and name != "conn_lost_cnt" and \
+               name != "lock_lost_cnt":
+                continue
+
+            stat_cnt = counter_dict.get(name)
+            if stat_cnt is None:
+                stat_cnt = StatusCounter(name, int(val))
+
+            stat_cnt.cnt = int(val)
+            # TODO:
+            # If we detect a degraded state, we can throttle the path here.
+            self.check_for_degraded_state(stat_cnt)
+            stat_cnt.last_cnt = stat_cnt.cnt
+
+            counter_dict[name] = stat_cnt
+
+        self.gw_counter_lookup[gw] = counter_dict
+
+
+class DeviceStatusWatcher(threading.Thread):
+    def __init__(self, logger):
+        threading.Thread.__init__(self)
+        self.logger = logger
+        self.daemon = True
+        self.cluster = None
+        self.status_lookup = {}
+
+    def get_dev_status(self, image_name):
+        return self.status_lookup.get(image_name)
+
+    def exit(self):
+        if self.cluster:
+            self.cluster.shutdown()
+
+    def run(self):
+        self.cluster = rados.Rados(conffile=settings.config.cephconf,
+                                   name=settings.config.cluster_client_name)
+        self.cluster.connect()
+
+        while True:
+            time.sleep(settings.config.status_check_interval)
+
+            cmd = json.dumps({"prefix": "service status", "format": "json"})
+
+            ret, outb, outs = self.cluster.mon_command(cmd, b'')
+            if ret != 0:
+                self.logger.error("mon command failed {}".format(ret))
+                continue
+
+            svc = json.loads(outb).get('tcmu-runner')
+            if svc is None:
+                continue
```

Comment:
Nit: can we track in the state that we have no data available from `tcmu-runner`? That might indicate an issue

Reply:
Yeah, this makes sense.

Reply:
BTW, when the `tcmu-runner` is not avaliable, the `status` output will be `{}`:

```
{
    "allocating_host": "lxbceph2",
    "backstore": "user:rbd",
    "backstore_object_name": "data.block01",
    "controls": {},
    "created": "2020/08/13 01:01:40",
    "image": "block01",
    "owner": "lxbceph1",
    "pool": "data",
    "pool_id": 9,
    "status": {},
    "updated": "2020/08/13 01:14:29",
    "wwn": "ad1a0c5c-a666-434b-90ab-dfc851ba5a02"
}
```

Will this be enough ?

Reply:
Yeah, could output `Unknown` or similar on the CLI

Reply:
Sure, will fix it. Thanks.

Reply:
Sure, will fix it.